### DISCUSSION TITLE
Upload Attachment via Public API

### DISCUSSION THREAD

**Discussion by ecxs** [Budibase/budibase#9168] [Help]
Is it possible to upload an attachment via the public API? And if so, how?

**Comment by thomaskamalakis** [comment]
Hi ecxs

What might work is exposing port 9000 of the minio container in the docker-compose.yaml and using the MINIO API to upload the file. This takes care of putting the file in the storage though, if you need to create a table entry with an attachment column pointing to this attachment you need to use the Budibase API to create the row, taking care to come up with the correct attachment column information.

**Reply by ecxs** [reply]
Hi @thomaskamalakis thanks for the tip! I was actually already looking at this, and having been playing around with the build in CouchDB and MinIO services. I managed to get an insight in how BudiBase organises their data with the help of Fauxton, for example:

![Screenshot 2022-12-30 at 15 27 18](https://user-images.githubusercontent.com/4005405/210081017-c94a763a-bf95-4d4e-bfd3-201d8435082d.png)

I can also upload files into the prod-budi-app-assets bucket, but can not see (yet) how/where to get the right path/url. This is what I get back from the S3 node in n8n when I search the prod-budi-app-assets bucket:

![Screenshot 2022-12-30 at 15 33 04](https://user-images.githubusercontent.com/4005405/210081528-f547d8e9-331a-4a6b-b2cd-38062f51e2a4.png)

I'm now going to try to find out how things are structured in MinIO ...

And to the guys and girls behind BudiBase: why are the attachments not stored in CouchDB ?

**Comment by thomaskamalakis** [comment]
Dear @ecxs 
I think the url has a part that depends on the bucket and app-id and another part which seems to be generated randomly together with the file extension. This must be in order to avoid conflicts of the type: two users upload the filename "image.png" to the same table, etc.

It looks like Budibase organizes uploaded files to something like 

```
app_id + '/attachments/' + filename
```
and the url is something like
```
'/' + bucket + '/' + app_id + '/attachments/' + filename
```
where filename is generated randomly plus the original file extension. This is my intuition though not exactly glanced at the source code.

Also not sure exactly how MINIO works, but I think one should upload the file using its API - maybe just putting the file in the container volume in the right path is not enough...

**Reply by ecxs** [reply]
Hi @thomaskamalakis , I think you might be right. I haven't figured out the used logic yet, and I think that instead of going down this road, I'm going to do a little write up of what I'm trying to achieve. Perhaps somebody has a better idea ;-)

**Reply by ArchonMegalon** [reply]
I can't believe, this is still not solved. I mean, storing something as binary is really a very common use case... :(

**Comment by NicolasFabre81** [comment]
Related to my question here https://github.com/Budibase/budibase/discussions/14645
